unison-share-api/src/Unison/Server/Share/FuzzyFind.hs

```diff
@@ -0,0 +1,358 @@
+        ( [ ( PathSegments,
+              Alignment,
+              UnisonName,
+              [Backend.FoundRef]
+            )
+          ]
+        )
+      alignments =
+        (preparedTerms <> preparedTypes)
+          & List.sortOn (\(_, Alignment {score}, _, _) -> score)
+  lift (join <$> traverse (loadEntry includeDependencies bh namesPerspective) alignments)
+  where
+    preparedQuery = prepareQuery (Text.unpack query)
+    limit = fromMaybe 10 mayLimit
+    loadEntry :: Bool -> BranchHash -> SqliteOps.NamesPerspective -> (PathSegments, Alignment, Text, [Backend.FoundRef]) -> IO [(Alignment, FoundResult)]
+    loadEntry includeDependencies bh searchPerspective (pathToMatch, a, n, refs) = do
+      namesPerspective <-
+        -- If we're including dependencies we need to ensure each match's type signature is
+        -- rendered using a ppe with that dependency's names.
+        -- So we re-compute the perspective for each match.
+        --
+        -- If not we can use the same perspective for every match.
+        if includeDependencies
+          then Codebase.runTransaction codebase $ SqliteOps.namesPerspectiveForRootAndPath bh (coerce (Path.toList perspective) <> pathToMatch)
+          else pure searchPerspective
+      let relativeToBranch = Nothing
+      entries <- for refs $
+        \case
+          Backend.FoundTermRef r -> do
+            Left . (r,) <$> Backend.termListEntry codebase relativeToBranch (ExactName (NameSegment n) (Cv.referent1to2 r))
+          Backend.FoundTypeRef r ->
+            Codebase.runTransaction codebase do
+              Right . (r,) <$> Backend.typeListEntry codebase relativeToBranch (ExactName (NameSegment n) r)
+      let allLabeledDependencies = foldMap (either (termEntryLabeledDependencies . snd) (typeEntryLabeledDependencies . snd)) entries
+      pped <- liftIO . Codebase.runTransaction codebase $ PPED.ppedForReferences namesPerspective allLabeledDependencies
+      let ppe = PPED.suffixifiedPPE pped
+      Codebase.runTransaction codebase do
+        for entries \case
+          Left (r, termEntry) ->
+            pure
+              ( a,
+                FoundTermResult
+                  . FoundTerm
+                    (Backend.bestNameForTerm @Symbol ppe (mayDefaultWidth typeWidth) r)
+                  $ Backend.termEntryToNamedTerm ppe typeWidth termEntry
+              )
+          Right (r, typeEntry) -> do
+            let namedType = Backend.typeEntryToNamedType typeEntry
+            let typeName = Backend.bestNameForType @Symbol ppe (mayDefaultWidth typeWidth) r
+            typeHeader <- Backend.typeDeclHeader codebase ppe r
+            let ft = FoundType typeName typeHeader namedType
+            pure (a, FoundTypeResult ft)
+
+-- Scores a matched name by the number of segments.
+-- Lower is better.
+scoreMatch :: S.NamedRef r -> Int
+scoreMatch S.NamedRef {S.reversedSegments = NameLookups.ReversedName segments} = length segments
+
+data Alignment = Alignment
```

Comment:
The frontend expects a structure similar to the FZF library, but we're not using that library anymore, so I just define the same wrapper types manually.

Reply:
Does some import go away?

Reply:
Ok, I see it.